### PR TITLE
[ML] Initialize incremental training

### LINKED ISSUE #1829: [ML] Initiate model for incremental training
In order to update the boosted tree model for incremental training, we need to be able to 
* Write training data summary into as a new type of results
* Use this data summary and the inference model definition to initialize a new model

Subtasks:
- [ ] Implement writing data summary as a new result type
- [ ] Implement a new input stream to pass compressed model definition and data summary to the `data_frame_analyzer`
- [ ] Implement a new method in `CBoostedTreeFactory` to create boosted trees from these data

### LINKED ISSUE #1851: [ML] Extend CDataSummarizationJsonSerializer::bestForestFromJsonStream to the vector valued leaves
At the moment we assume that node leaves have only 1-element vectors. This functionality has to be extended to support multiclass classification as well.

https://github.com/valeriy42/ml-cpp/blob/4619583f79134c92e3fb67cb9b899795c4a43e17/lib/api/CDataSummarizationJsonSerializer.cc#L301-L301

### PR DESCRIPTION
**Pull request by valeriy42**
This PR implements initialization of the boosted tree object with pre-trained best forest and data summarization.

The main control points happen in the constructor `CDataFrameTrainBoostedTreeRunner::CDataFrameTrainBoostedTreeRunner` and in the new method`CBoostedTreeFactory::constructFromDefinition`. I added a new `TASK` parameter to the runner configuration, which be `train` or `update` and will call different factory demthods depending on the task value.

I added boilerplate code for the test `CDataFrameAnalyzerTrainingTest/testRunBoostedTreeRegressionIncrementalTraining`. The test doesn't have any checks so far because incremental training is not wired in yet. Hence, I disabled that test for now.

I added `CDataFrameTrainBoostedTreeRegressionRunnerTest//testCreationForIncrementalTraining` to test  correct initialization of data summarization and the best forest.

Closes #1851.
Closes #1829.

### CODE REVIEW

**Review by tveasey — COMMENTED** [review]
I did a pass through. I'd like to have another read through on Monday as well. I think my main high-level comment is I'd very much like to avoid creating a temporary data frame and write straight into the single copy we've had until now, but I think we can postpone this until we've joined the dots. I had a few smaller comments mainly around efficiency and a few places I think we can improve efficiency.

File: include/maths/CBoostedTree.h [hunk 1]
```diff
@@ -114,6 +114,9 @@ class MATHS_EXPORT CBoostedTreeNode final {
     //! Set the number of samples to \p value.
     void numberSamples(std::size_t value);
 
+    //! Set the node value to \p nodeValue.
+    void nodeValue(TVector nodeValue);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTree.h:118]: You are right. I missed it. :+1:

File: include/maths/CBoostedTreeFactory.h [hunk 2]
```diff
@@ -62,6 +74,13 @@ class MATHS_EXPORT CBoostedTreeFactory final {
     //! \warning Throws runtime error on fail to restore.
     static CBoostedTreeFactory constructFromString(std::istream& jsonStream);
 
+    static CBoostedTreeFactory
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTreeFactory.h:77]: Description added.

File: include/maths/CBoostedTreeFactory.h [hunk 3]
```diff
@@ -42,6 +43,17 @@ class MATHS_EXPORT CBoostedTreeFactory final {
     using TTrainingStateCallback = CBoostedTree::TTrainingStateCallback;
     using TLossFunctionUPtr = CBoostedTree::TLossFunctionUPtr;
     using TAnalysisInstrumentationPtr = CDataFrameAnalysisInstrumentationInterface*;
+    using TDataFrameUPtr = std::unique_ptr<core::CDataFrame>;
+    using TEncoderUPtr = std::unique_ptr<maths::CDataFrameCategoryEncoder>;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTreeFactory.h:47]: Fixed

File: lib/api/CDataFrameTrainBoostedTreeRunner.cc [hunk 4]
```diff
@@ -145,9 +150,31 @@ CDataFrameTrainBoostedTreeRunner::CDataFrameTrainBoostedTreeRunner(
...
-    m_BoostedTreeFactory = std::make_unique<maths::CBoostedTreeFactory>(
-        maths::CBoostedTreeFactory::constructFromParameters(
-            this->spec().numberThreads(), std::move(loss)));
+    switch (m_Task) {
+    case (E_Train):
+        m_BoostedTreeFactory = std::make_unique<maths::CBoostedTreeFactory>(
+            maths::CBoostedTreeFactory::constructFromParameters(
+                this->spec().numberThreads(), std::move(loss)));
+        break;
+    case (E_Update):
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataFrameTrainBoostedTreeRunner.cc:159]: Fixed

[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataFrameTrainBoostedTreeRunner.cc:159]: Fixed

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 5]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+    rapidjson::IStreamWrapper isw(*istream);
+    rapidjson::Document doc;
+    doc.ParseStream(isw);
+    if (doc.HasMember(JSON_COMPRESSED_DATA_SUMMARIZATION_TAG) &&
+        doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG].IsObject()) {
+        auto& compressedDataSummarization = doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG];
+        if (compressedDataSummarization.HasMember(JSON_DATA_SUMMARIZATION_TAG) &&
+            compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].IsString()) {
+            std::stringstream compressedStream{
+                compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].GetString()};
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:153]: Nice. Thank you for the idea!

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 6]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+}
+
+CRetrainableModelJsonDeserializer::TBestForest
+CRetrainableModelJsonDeserializer::bestForestFromJsonStream(const core::CDataSearcher::TIStreamP& istream) {
+    using TNodeVec = maths::CBoostedTreeFactory::TNodeVec;
+    using TNodeVecVec = maths::CBoostedTreeFactory::TNodeVecVec;
+    rapidjson::IStreamWrapper isw(*istream);
+    rapidjson::Document d;
+    d.ParseStream(isw);
+    if (d.HasMember(CInferenceModelDefinition::JSON_TRAINED_MODEL_TAG) &&
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:280]: Good idea. Added error logging and removed nesting.

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 7]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+    if (doc.HasMember(JSON_COMPRESSED_DATA_SUMMARIZATION_TAG) &&
+        doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG].IsObject()) {
+        auto& compressedDataSummarization = doc[JSON_COMPRESSED_DATA_SUMMARIZATION_TAG];
+        if (compressedDataSummarization.HasMember(JSON_DATA_SUMMARIZATION_TAG) &&
+            compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].IsString()) {
+            std::stringstream compressedStream{
+                compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].GetString()};
+            auto decompressedSPtr =
+                std::make_shared<std::stringstream>(decompressStream(std::stringstream(
+                    compressedDataSummarization[JSON_DATA_SUMMARIZATION_TAG].GetString())));
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:156]: It was actually refactored out. I changed the code and remove the variable.

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 8]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+        d[CInferenceModelDefinition::JSON_TRAINED_MODEL_TAG].IsObject()) {
+        auto trainedModel = d[CInferenceModelDefinition::JSON_TRAINED_MODEL_TAG].GetObject();
+        if (trainedModel.HasMember(CEnsemble::JSON_ENSEMBLE_TAG) &&
+            trainedModel[CEnsemble::JSON_ENSEMBLE_TAG].IsObject()) {
+            auto ensemble = trainedModel[CEnsemble::JSON_ENSEMBLE_TAG].GetObject();
+            if (ensemble.HasMember(CEnsemble::JSON_TRAINED_MODELS_TAG) &&
+                ensemble[CEnsemble::JSON_TRAINED_MODELS_TAG].IsArray()) {
+                auto trainedModels =
+                    ensemble[CEnsemble::JSON_TRAINED_MODELS_TAG].GetArray();
+                auto forest = std::make_unique<TNodeVecVec>();
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:290]: Fixed

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 9]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+                            // TODO #1851 this can be/is a vector
+                            maths::CBoostedTreeNode::TVector nodeValue(1);
+                            nodeValue[0] =
+                                node[CTree::CTreeNode::JSON_LEAF_VALUE_TAG].GetDouble();
+                            nodes[nodeIndex].numberSamples(numberSamples);
+                            nodes[nodeIndex].nodeValue({nodeValue});
+
+                        } else {
+                            std::size_t splitFeature{
+                                node[CTree::CTreeNode::JSON_SPLIT_FEATURE_TAG].GetUint64()};
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:310]: Good point. I added accessors and fallback values for error handling.

File: lib/api/CDataSummarizationJsonSerializer.cc [hunk 10]
```diff
@@ -87,5 +138,225 @@ void CDataSummarizationJsonSerializer::addToJsonStream(TGenericLineWriter& write
...
+    return nullptr;
+}
+
+CRetrainableModelJsonDeserializer::TBestForest
+CRetrainableModelJsonDeserializer::bestForestFromDocumentCompressed(
+    const core::CDataSearcher::TIStreamP& istream) {
+    rapidjson::IStreamWrapper isw(*istream);
+    rapidjson::Document d;
+    d.ParseStream(isw);
+    if (d.HasMember(CInferenceModelDefinition::JSON_COMPRESSED_INFERENCE_MODEL_TAG) &&
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/api/CDataSummarizationJsonSerializer.cc:337]: Fixed

**Comment by valeriy42** [comment]
Thank you for your review @tveasey. I addressed your comments. It would be great if you could take another look.

**Comment by tveasey** [comment]
Great thanks! Reviewing now.

**Review by tveasey — APPROVED** [review]
Just a request to add a TODO comment, but LGTM.

File: include/maths/CBoostedTreeFactory.h [hunk 11]
```diff
@@ -143,6 +169,12 @@ class MATHS_EXPORT CBoostedTreeFactory final {
     //! \warning A tree object can only be restored once.
     TBoostedTreeUPtr restoreFor(core::CDataFrame& frame, std::size_t dependentVariable);
 
+    //! \name Test Only
+    //@{
+    //! Get the boosted tree implementation.
+    const CBoostedTreeImpl& boostedTreeImpl() const;
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply include/maths/CBoostedTreeFactory.h:177]: Added todo in d021c22

**Pull request merged**